src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2523,6 +2521,59 @@ private bool Fix(int iParam, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
             }
         }
 
+        /// <summary>
+        /// Takes the dynamic flags from both types and applies them onto the target.
+        /// </summary>
+        internal static TypeSymbol MergeDynamic(TypeSymbol first, TypeSymbol second, TypeSymbol target, Compilation compilation)
+        {
+            // SPEC: 4.7 The Dynamic Type
+            //       Type inference (7.5.2) will prefer dynamic over object if both are candidates.
+            if (first.Equals(second, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds | TypeCompareKind.IgnoreDynamic))
+            {
+                return target;
+            }
+            ImmutableArray<bool> flags1 = CSharpCompilation.DynamicTransformsEncoder.EncodeWithoutCustomModifierFlags(first, RefKind.None);
+            ImmutableArray<bool> flags2 = CSharpCompilation.DynamicTransformsEncoder.EncodeWithoutCustomModifierFlags(second, RefKind.None);
+            Debug.Assert(flags1.Length == flags2.Length);
+            ImmutableArray<bool> mergedFlags = flags1.SelectAsArray((f, index, other) => f | other[index], flags2);
+
+            return DynamicTypeDecoder.TransformTypeWithoutCustomModifierFlags(target, (AssemblySymbol)compilation.Assembly, RefKind.None, mergedFlags);
+        }
+
+        /// <summary>
+        /// Takes the names from the two types, finds the common names, and applies them onto the target.
+        /// </summary>
+        internal static TypeSymbol MergeTupleNames(TypeSymbol first, TypeSymbol second, TypeSymbol target, Compilation compilation)
+        {
+            if (!target.ContainsTuple() ||
+                first.Equals(second, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds | TypeCompareKind.IgnoreDynamic) ||
+                !target.ContainsTupleNames())
+            {
+                return target;
+            }
+
+            ImmutableArray<string> names1 = CSharpCompilation.TupleNamesEncoder.Encode(first);
+            ImmutableArray<string> names2 = CSharpCompilation.TupleNamesEncoder.Encode(second);
+
+            ImmutableArray<string> mergedNames;
+            if (names2.IsDefault)
+            {
+                mergedNames = default(ImmutableArray<string>);
+            }
+            else
+            {
+                Debug.Assert(names1.Length == names2.Length);
+                mergedNames = names1.SelectAsArray((s, index, other) => string.CompareOrdinal(s, other[index]) == 0 ? s : null, names2);
```

Comment:
Yes, it is possible for either `names1` or `names2` to be default (if either side has no tuple names, then the `Encode` method returns a default array).

No, the `target` parameter cannot be removed. It is the merged type that is being worked on and it can be different from `first` or `second`, at least for `MergeTupleNames` (which is invoked after `MergeDynamic`). 
I didn't want the code to be reliant on this call order assumption, so I gave both `Merge` functions the same signatures (they can be called in either order safely).
Does that make sense?